app/eth2wrap/eth2wrap.go

```diff
@@ -228,6 +230,42 @@ func isSyncingError(err error) bool {
 	return strings.Contains(msg, "syncing")
 }
 
+// isBadGateway returns true when the error indicates a connectivity or upstream gateway issue.
+func isBadGateway(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	for current := err; current != nil; current = errors.Unwrap(current) {
```

Comment:
Good! A bit less readable as you pointed out, but better than matching random strings.